src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java

```diff
@@ -0,0 +1,113 @@
+
+import java.util.Arrays;
+
+import org.agrona.DirectBuffer;
+import org.agrona.MutableDirectBuffer;
+import org.agrona.collections.Hashing;
+import org.agrona.concurrent.UnsafeBuffer;
+
+/**
+ * A chunk of shared memory for storing incomplete data for decoding purposes.
+ * Methods are provided for getting a buffer and appending partial data in the case of an
+ * incomplete block of data to be decoded. The memory is segmented into slots, the size of
+ * each slot being the maximum allowed size for the header portion of an HTTP request or response.
+ * <b>Each instance of this class is assumed to be used by one and only one thread.</b>
+ */
+class Slab
+{
+    static final int OUT_OF_MEMORY = -2;
+    private static final int UNUSED = -1;
+    private final int slotSize;
+    private final int slotSizeBits;
+    private final MutableDirectBuffer buffer;
+    private final int offsets[]; // starting position for slot data in buffer
+    private final int remaining[]; // length of data in slot
+    private int acquiredSlots = 0;
+
+    Slab(int minimumNumberOfSlots, int minimumSlotSize)
+    {
+        int slots = findNextPositivePowerOfTwo(minimumNumberOfSlots);
+        slotSize = findNextPositivePowerOfTwo(minimumSlotSize);
+        slotSizeBits = Integer.numberOfTrailingZeros(slotSize);
+        int capacityInBytes = slots << slotSizeBits;
+        buffer = new UnsafeBuffer(new byte[capacityInBytes]);
+        offsets = new int[slots];
+        Arrays.fill(offsets, UNUSED);
+        remaining = new int[slots];
+    }
+
+    public DirectBuffer buffer(int slot)
+    {
+        return buffer;
+    }
+
+    public int offset(int slot)
+    {
+        return offsets[slot];
+    }
+
+    public int limit(int slot)
+    {
+        return offsets[slot] + remaining[slot];
+    }
+
+    public int append(int slot, DirectBuffer written, int offset, int limit)
+    {
+        assert slot >= 0 && slot < offsets.length;
+        assert offsets[slot] != UNUSED;
+        assert limit > offset;
+        int dataLength = limit - offset;
+        if (dataLength + remaining[slot] >= slotSize)
+        {
+            free(slot);
+            return OUT_OF_MEMORY;
+        }
+        buffer.putBytes(offsets[slot] + remaining[slot], written, offset, dataLength);
+        remaining[slot] += dataLength;
+        return slot;
+    }
+
+    public int acquire(long streamId)
+    {
+        if (acquiredSlots == offsets.length)
+        {
+            return OUT_OF_MEMORY;
+        }
+        final int mask = offsets.length - 1;
+        int slot = Hashing.hash(streamId, mask);
+        while (offsets[slot] != UNUSED)
+        {
+            slot = ++slot & mask;
+        }
+        offsets[slot] = slot << slotSizeBits;
+        acquiredSlots++;
+        return slot;
+    }
+
+    public void free(int slot)
+    {
+        offsets[slot] = UNUSED;
+        remaining[slot] = 0;
+        acquiredSlots--;
+    }
+
+}
+
```

Comment:
As discussed, let's modify the above to use a `BitSet` to track available slots, and leave mutability outside the `Slab`.